### PR TITLE
[data] [streaming] Support iterating over streaming_split() iterators multiple times

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

In order to support Train integration, the iterators returned from `Dataset.streaming_split()` must be repeatable, as with the other iterators.

Split out from https://github.com/ray-project/ray/pull/33276

### CODE REVIEW

File: doc/source/ray-observability/ray-logging.rst [hunk 1]
```diff
@@ -95,6 +95,8 @@ Limitations:
 - Only a subset of tqdm functionality is supported. Refer to the ray_tqdm `implementation <https://github.com/ray-project/ray/blob/master/python/ray/experimental/tqdm_ray.py>`__ for more details.
 - Performance may be poor if there are more than a couple thousand updates per second (updates are not batched).
 
+Tip: To avoid `print` statements from the driver conflicting with tqdm output, use `ray.experimental.tqdm_ray.safe_print` instead.
```
Comment (ericl) [inline doc/source/ray-observability/ray-logging.rst:98]: Some incidental fixes.

File: python/ray/data/_internal/progress_bar.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ class ProgressBar:
 
     def __init__(self, name: str, total: int, position: int = 0):
         self._desc = name
-        if not _enabled or threading.current_thread() is not threading.main_thread():
+        if not _enabled:
```
Comment (ericl) [inline python/ray/data/_internal/progress_bar.py:49]: No longer needed with the distributed tqdm change.

File: python/ray/data/tests/test_streaming_integration.py [hunk 3]
```diff
@@ -109,35 +123,39 @@ def get_lengths(*iterators, use_iter_batches=True):
...
-        if equal_split:
-            assert lengths == [333, 333, 333], lengths
-        else:
-            assert lengths == [300, 300, 400], lengths
+        for _ in range(2):
+            lengths = get_lengths(i1, i2, i3, use_iter_batches=use_iter_batches)
+            if equal_split:
+                assert lengths == [333, 333, 333], lengths
+            else:
+                assert lengths == [300, 300, 400], lengths
```
Comment (amogkam) [inline python/ray/data/tests/test_streaming_integration.py:158]: add a test for the barrier case? simulating one consumer iterating through the  epoch faster than the other.

**Review by amogkam — COMMENTED** [review]
lgtm, but would be great if we can also add a test for the barrier case.

File: python/ray/data/tests/test_streaming_integration.py (hunk 3 above)
Comment (ericl) [reply python/ray/data/tests/test_streaming_integration.py:158]: Good suggestion. There is actually a bug here if iterators are copied.

**Comment by ericl** [comment]
@c21 @jianoaix stamp?

File: python/ray/data/_internal/stream_split_dataset_iterator.py [hunk 4]
```diff
@@ -201,3 +236,37 @@ def get(self, output_split_idx: int) -> Optional[ObjectRef[Block]]:
...
+        """Arrive and block until the start of the given epoch."""
+
+        # Decrement and await all clients to arrive here.
+        with self._lock:
+            starting_epoch = self._cur_epoch
+            self._unfinished_clients_in_epoch -= 1
+
+        start_time = time.time()
+        while (
+            self._cur_epoch == starting_epoch and self._unfinished_clients_in_epoch != 0
```
Comment (jianoaix) [inline python/ray/data/_internal/stream_split_dataset_iterator.py:250]: Does the read of these two variables need to be protected by reader lock?

**Review by jianoaix — APPROVED** [review]

**Comment by jianoaix** [comment]
This effectively supports epoch now: https://github.com/ray-project/ray/issues/33205

File: python/ray/data/_internal/stream_split_dataset_iterator.py (hunk 4 above)
Comment (ericl) [reply python/ray/data/_internal/stream_split_dataset_iterator.py:250]: These should be effectively atomic accesses, so it's safe.

**Review by c21 — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Pull request merged**